test/test_util_pipe.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+import concurrent.futures
+from os.path import join
+
+from test.picardtestcase import PicardTestCase
+
+from picard import (
+    PICARD_APP_NAME,
+    PICARD_FANCY_VERSION_STR,
+)
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+from picard.util import pipe
+
+
+def pipe_listener(pipe_handler, end_of_sequence):
+    IGNORED_OUTPUT = {pipe.Pipe.MESSAGE_TO_IGNORE, pipe.Pipe.NO_RESPONSE_MESSAGE, "", end_of_sequence}
+    received = []
+    messages = []
+
+    while end_of_sequence not in messages:
+        messages = pipe_handler.read_from_pipe()
+        for message in messages:
+            if message not in IGNORED_OUTPUT:
+                received.append(message)
+
+    return tuple(received)
+
+
+def pipe_writer(pipe_handler, to_send, end_of_sequence):
+    for message in to_send:
+        while not pipe_handler.send_to_pipe(message):
+            pass
+    while not pipe_handler.send_to_pipe(end_of_sequence):
+        pass
+
+
+class TestPipe(PicardTestCase):
+
+    def test_invalid_args(self):
+        # Pipe should be able to make args iterable (last argument)
+        self.assertRaises(pipe.PipeErrorInvalidArgs, pipe.Pipe, PICARD_APP_NAME, PICARD_FANCY_VERSION_STR, 1)
+        self.assertRaises(pipe.PipeErrorInvalidAppData, pipe.Pipe, 21, PICARD_FANCY_VERSION_STR, None)
+        self.assertRaises(pipe.PipeErrorInvalidAppData, pipe.Pipe, PICARD_APP_NAME, 21, None)
+
+    def test_filename_generation(self):
+        handler = pipe.Pipe(PICARD_APP_NAME, PICARD_FANCY_VERSION_STR)
+        suffix = f"{PICARD_APP_NAME}_v{PICARD_FANCY_VERSION_STR}_pipe_file"
+        MAC_PATH = join(pipe.Pipe.PIPE_MAC_DIR, suffix)
+        WIN_PATH = join(pipe.Pipe.PIPE_WIN_DIR, suffix.replace('.', '-'))
+
+        UNIX_PATHS = {
+            join(pipe.Pipe.PIPE_UNIX_DIR, suffix),
+            join(pipe.Pipe.PIPE_UNIX_FALLBACK_DIR, suffix)
+        }
+
+        if IS_MACOS:
+            self.assertEquals(handler.path, MAC_PATH,
+                              "Pipe path generation")
+        elif IS_WIN:
```

Comment:
Instead you can run different tests for each platform.
See, for example, https://github.com/metabrainz/picard/blob/3b26fe8c425743083a96f9a045c388bafb9ccd96/test/test_file.py#L291

Something like:

```python
@unittest.skipUnless(IS_WIN, "windows test")
def test_filename_generation_win(self):
    ...

@unittest.skipUnless(IS_MACOS, "macos test")
def test_filename_generation_macos(self):
    ...

@unittest.skipUnless(not IS_MACOS and not IS_WIN, "non-macos/win test")
def test_filename_generation(self):
    ...

```

Check other test files, there are various places where we do that.
You can move common code to a common method to avoid code duplication.

Reply:
The issue with tests is that paths defined in https://github.com/metabrainz/picard/blob/5f27781635610c8564da1c2ad4de3dbb6571ec3d/picard/util/pipe.py#L94-L97 aren't all defined in the context of tests.
You need to create a temporary dir for tests, and override those paths. For example, there's no guarantee the user tests run as has a home dir. You also need to define env variables that may affect tests (XDG_*) and "simulate" a specific env.

Reply:
`os.getenv()` doesn't require its args to be defined, so we can keep it that way. It'll be just `None` and current checks are okay.

But we should move `os.path.expanduser()` to the filename generation part

Reply:
`os.environ.get()` allows to set what should be returned when no such variable was found and this should do: to replace `$HOME` with `.`

Reply:
Imo `/tmp` is a good fallback. A user without `$HOME` and `$XDG_RUNTIME_DIR` is highly unlikely to be using Picard.